### PR TITLE
Create strategies for strings matching Lark (EBNF) context-free grammars

### PR DESCRIPTION
**Pull request by Zac-HD**
Being an initial implementation of #170, based on the [Lark parser package](https://github.com/lark-parser/lark/).

### CODE REVIEW

File: hypothesis-python/RELEASE.rst [hunk 1]
```diff
@@ -0,0 +1,7 @@
+RELEASE_TYPE: minor
+
+This release adds a new module, ``hypothesis.extra.grammar``, which you
```
Comment (DRMacIver) [inline hypothesis-python/RELEASE.rst]: I think maybe we want to think about whether we really want a single `grammar` extra namespace here - we're likely to want to support other things and we don't want to force people to depend on all of them just to use one of them.

**Comment by Zac-HD** [comment]
I can do more testing, easy :smile: 

--------------

Re: namespaces, I spent quite a while looking for Python libraries that would give us nice data structures representing ABNF or ANTLR grammars.  I didn't find any, hence using Lark here.

The obvious future direction (at least to me) is to parse more common grammar formats *using Lark*, and run a transformer on the parse tree that will give us a strategy directly.  [Here's an initial proof of concept](https://gist.github.com/Zac-HD/b44d32221fd5a28f67d7c2965e54855c), from a few evenings' experimentation.  It's obviously early days, but I'm fairly confident that we won't need other dependencies to support other grammars.

In the worst case, we can just ensure that it's possible to use `hypothesis.extra.grammar.from_abnf` (for e.g.) without the dependencies for other grammar types.  So `pip install hypothesis[grammar]` would get you lark-parser and some-abnf-library, but `pip install hypothesis lark-parser` would be enough to use `from_lark`.  Or don't bother at all, as lark-parser has no dependencies so it doesn't matter that much.

All that said, if you still think a `grammar` namespace is too general we can call this `extra.lark`

**Comment by DRMacIver** [comment]
> All that said, if you still think a grammar namespace is too general we can call this extra.lark

I think my preference would be one of:

* `hypothesis.extra.lark`
* `hypothesis.grammars` (no extra) which is importable without lark but `from_lark` errors if lark is not installed.

having it in the extra model where it's tied to one of multiple versions is a bit weird.

Also, especially if we're planning to use this for other grammars and/or want to use this as the basis of other strategies (which I feel would be a good idea), we *could* just move this into the main namespace and add an unconditional dependency on `lark`?

**Comment by Zac-HD** [comment]
"Just add a hard dependency" is attractive, but I'd prefer to stay minimal for now so I'll convert to `extra.lark`.  We can always move it to core later, as we did for date and time strategies!

**Comment by Zac-HD** [comment]
@DRMacIver - moved to `extra.lark`, and additional tests as requested.

**Pull request merged**